cpp/src/randomforest/randomforest.cu

```diff
@@ -340,14 +393,211 @@ RF_metrics rfClassifier<T>::cross_validate(const cumlHandle& user_handle, const
+				n_sampled_rows, 0, 8*sizeof(unsigned int), stream));
+		} else { // Sampling w/o replacement
+			MLCommon::device_buffer<unsigned int> *inkeys = new MLCommon::device_buffer<unsigned int>(handle.getDeviceAllocator(), stream, n_rows);
+			MLCommon::device_buffer<unsigned int> *outkeys = new MLCommon::device_buffer<unsigned int>(handle.getDeviceAllocator(), stream, n_rows);
+			thrust::sequence(thrust::cuda::par.on(stream), inkeys->data(), inkeys->data() + n_rows);
+			int *perms = nullptr;
+			MLCommon::Random::permute(perms, outkeys->data(), inkeys->data(), 1, n_rows, false, stream);
+			// outkeys has more rows than selected_rows; doing the shuffling before the resize to differentiate the per-tree rows sample.
+			CUDA_CHECK(cub::DeviceRadixSort::SortKeys((void *)rows_temp_storage->data(), temp_storage_bytes, outkeys->data(), sorted_selected_rows.data(),
+				n_sampled_rows, 0, 8*sizeof(unsigned int), stream));
+			inkeys->release(stream);
+			outkeys->release(stream);
+			delete inkeys;
+			delete outkeys;
+		}
+
+		/* Build individual tree in the forest.
+		   - input is a pointer to orig data that have n_cols features and n_rows rows.
+		   - n_sampled_rows: # rows sampled for tree's bootstrap sample.
+		   - sorted_selected_rows: points to a list of row #s (w/ n_sampled_rows elements) used to build the bootstrapped sample.
+			Expectation: Each tree node will contain (a) # n_sampled_rows and (b) a pointer to a list of row numbers w.r.t original data.
+		*/
+		trees[i].fit(user_handle, input, n_cols, n_rows, labels, sorted_selected_rows.data(), n_sampled_rows, this->rf_params.tree_params);
+
+		//Cleanup
+		rows_temp_storage->release(stream);
+		selected_rows.release(stream);
+		sorted_selected_rows.release(stream);
+		delete rows_temp_storage;
+	}
+}
+
+/**
+ * @brief Predict target feature for input data; regression for single feature supported.
+ * @tparam T: data type for input data (float or double).
+ * @param[in] user_handle: cumlHandle (currently unused; API placeholder)
+ * @param[in] input: test data (n_rows samples, n_cols features) in row major format. CPU pointer.
+ * @param[in] n_rows: number of  data samples.
+ * @param[in] n_cols: number of features (excluding target feature).
+ * @param[in, out] predictions: n_rows predicted labels. CPU pointer, user allocated.
+ * @param[in] verbose: flag for debugging purposes.
+ */
+template<typename T>
+void rfRegressor<T>::predict(const cumlHandle& user_handle, const T * input, int n_rows, int n_cols, T * predictions, bool verbose) const {
+
+	ASSERT(trees, "Cannot predict! No trees in the forest.");
+	ASSERT((n_rows > 0), "Invalid n_rows %d", n_rows);
+	ASSERT((n_cols > 0), "Invalid n_cols %d", n_cols);
+	ASSERT(predictions != nullptr, "Error! User has not allocated memory for predictions.");
+
+	int row_size = n_cols;
+
+	for (int row_id = 0; row_id < n_rows; row_id++) {
```

Comment:
Doing this inference in serial seems like quite a performance bottleneck as the data sizes grow. Assuming the call to `trees[i].predict` is read-only, it looks like each loop iteration is accessing their own unique entry of the `predictions` array. Have you looked into parallelizing this loop?

Reply:
As per the Developer Guide https://github.com/rapidsai/cuml/blob/branch-0.8/cpp/DEVELOPER_GUIDE.md we need to ensure thread safety.

Hence we keep it this way for now and later move inference on the GPU.

Reply:
I have not started the public discussion about the `cuML is single threaded` statement yet, but we need to have it, as well as explicit guidelines for its purpose and meaning, whilst I propose it be removed altogether and replaced with a better description of the expected behavior.

Adding threads to process host memory in parallel does not jeopardize the thread-safety of an algorithm. Thread safety becomes jeopardized when the exact same region of memory is updated in multiple threads without proper synchronization. In this case, I don't believe you would need any locking since the `row_id` is manipulating different regions of memory. If you did need locking, however, the algorithm would still be thread-safe.

Since Jiri is gone for the time being, I briefly discussed the `cuML is single threaded` statement with @teju85 to get a better understanding. According to him, so long as omp can be turned off in the build, which it can, we are okay to use it for cases like this. 

From what I've been told, Jiri has a customer that needs this turned off, but that does not seem to be a compelling reason for slowing down the inference for all other users of cuML.

Reply:
I realized my comment above is more related to the wording of the developer guide than it is to the RF predict function. Given that we intend to do RF inference on device by 0.9, I certainly have no problems keeping this serial loop on host until then.

Reply:
Agreed for GPU based RF inference by 0.9

Reply:
> I have not started the public discussion about the `cuML is single threaded` statement yet, but we need to have it, as well as explicit guidelines for its purpose and meaning, whilst I propose it be removed altogether and replaced with a better description of the expected behavior.
> 
> Adding threads to process host memory in parallel does not jeopardize the thread-safety of an algorithm. Thread safety becomes jeopardized when the exact same region of memory is updated in multiple threads without proper synchronization. In this case, I don't believe you would need any locking since the `row_id` is manipulating different regions of memory. If you did need locking, however, the algorithm would still be thread-safe.
> 
> I briefly discussed the `cuML is single threaded` statement with @teju85 to get a better understanding. According to him, so long as `omp` can be turned off in the build, which it can, we are okay to use it for cases like this. Tagging @jirikraus.

I agree that the `cuML is single threaded` statement needs more discussion and clarification. I started the developer guide with this bold statement to trigger that discussion, which happened now :-). The rational for this restrictive statement is that adding threads (and additional dependencies like an OpenMP runtime) can cause issues in some cases so it is better to use a single thread per default and only use multiple threads where they really provide benefit (+provide an option to turn off multi threaded and external dependencies like OpenMP). Does that make sense?